File: MATLAB/anovas/stats_anova_2way_mixed_ANOVA.m
%%
%   COURSE: Master statistics and machine learning: intuition, math, code										
%      URL: udemy.com/course/statsml_x/?couponCode=202304 
% 
%  SECTION: Analysis of Variance (ANOVA)
%    VIDEO: Two-way mixed-effects ANOVA
% 
%  TEACHER: Mike X Cohen, sincxpress.com
%

%%

% a clear MATLAB workspace is a clear mental workspace
close all; clear, clc

%% the data and group labels

data = csvread('TwoWayMixedANOVA_data.csv');

% correct way
group = [ repmat("Gr_A",15,1); repmat("Gr_B",15,1); repmat("Gr_C",15,1) ];


%%% note: the following is OK
%group = [ repmat('1',15,1); repmat('2',15,1); repmat('3',15,1) ];
% but the following is WRONG!!!!!
%group = [ repmat(1,15,1); repmat(2,15,1); repmat(3,15,1) ];


%% convert the data into a MATLAB table

tabl = table(group,data(:,1),data(:,2),data(:,3),'VariableNames',{'group';'t1';'t2';'t3'});

% correct way
repeatedFactor = table(["t1";"t2";"t3"],'VariableNames',{'TimePoints'});

% WRONG!!! (notice the numbers instead of char/string
% repeatedFactor = table((1:3)','VariableNames',{'TimePoints'});

%% run the ANOVA

% create the repeated-measures model
rm = fitrm(tabl,'t1-t3 ~ group','WithinDesign',repeatedFactor);

% ANOVA
ranova(rm,'WithinModel','TimePoints')

% multiple comparisons across groups
multcompare(rm,'TimePoints')

%% now to visualize the results

% hmmm...
plot(rm,'group','group')

% let's do this by organizing the data into a column with grouping variables
figure(2), clf
gr1 = repmat(1:3,size(data,1),1);
gr2 = [1*ones(15,1); 2*ones(15,1); 3*ones(15,1)];
gr2 = [gr2; gr2; gr2];

boxplot(data(:),[ gr1(:) gr2(:)],'colorgroup',gr1(:),'factorgap',[20 0])

%% done.

File: Python/anovas/stats_anova_2way_mixed_ANOVA.py
#!/usr/bin/env python
# coding: utf-8

# # COURSE: Master statistics and machine learning: Intuition, Math, code
# ##### COURSE URL: udemy.com/course/statsml_x/?couponCode=202304 
# ## SECTION: ANOVA
# ### VIDEO: Two-way mixed-effects ANOVA
# #### TEACHER: Mike X Cohen, sincxpress.com

# In[ ]:


# import libraries
import numpy as np
import matplotlib.pyplot as plt
import pingouin as pg
import pandas as pd
import seaborn as sns


# In[ ]:


## the data and group labels

data = np.loadtxt(open("TwoWayMixedANOVA_data.csv"),delimiter=",")

timepoint = ['1']*45 + ['2']*45 + ['3']*45
groups    = ['1']*15 + ['2']*15 + ['3']*15
s = []
for i in range(45):
    s += [str(i)]

# # convert to pandas
df = pd.DataFrame(data=np.matrix.flatten(data,'F'),columns=['TheData'])
df['Group'] = np.tile(groups,3)
df['TimePoint'] = timepoint
df['Subject'] = np.tile(s,3)

pd.set_option("display.max_rows", None, "display.max_columns", None)
df


# In[ ]:


pg.mixed_anova(data=df,dv='TheData',between='Group',within='TimePoint',subject='Subject')


# In[ ]:


sns.boxplot(data=df,hue="Group",y="TheData",x='TimePoint')

